### PR TITLE
[MLAS] Fix scratch buffer over-allocation in batched/grouped NCHW Conv

### LINKED ISSUE #28466: [Performance] Speed-vs-memory regression on Conv3D FusedConv as of version 1.23.2
### Describe the issue

The peak CPU virtual memory allocation of some of our models went from 8 GB to 71 GB with onnxruntime-gpu change from version 1.23.0 to 1.23.2. This large allocation continues to happen on the latest version tested 1.25.1. This happens on both Windows and Linux, using the CPU execution provider.

Here is the message when the allocation fails the large allocation: 
onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running FusedConv node. Name:'Conv__4386' Status Message: /onnxruntime_src/onnxruntime/core/framework/bfc_arena.cc:358 void* onnxruntime::BFCArena::AllocateRawInternal(size_t, bool, onnxruntime::Stream*) Failed to allocate memory for requested buffer of size 76672189440

It is possible to reduce the peak virtual memory usage using session options. Setting enable_cpu_mem_arena to False drops that peak alloc to ~40 GB. Changing intra_op_num_threads to 1 reduces that large alloc to the nominal value (~8 GB), however at that value the processing time is about doubled vs version 1.23.0. As intra_op_num_threads is increased the speed increases but so does the mem alloc size. So just using that session option I am unable to get back to the same speed-mem operating point, I am only able to choose between slower or larger virtual memory usage vs version 1.23.0.

### To reproduce

Here is a repro:
```
    tf.keras.backend.set_image_data_format("channels_last")
    x = tf.keras.Input((32, 64, 64, 1))
    m = tf.keras.Model(x, tf.keras.layers.Conv3D(6, 7, padding="same")(x))
    onnx.save(tf2onnx.convert.from_keras(m)[0], "{os.path.abspath(ONNX_FILE)}")

    sess = ort.InferenceSession("{os.path.abspath(ONNX_FILE)}", providers=["CPUExecutionProvider"])
    inp = sess.get_inputs()[0]
    sess.run(None, {{inp.name: np.random.randn(12, *inp.shape[1:]).astype(np.float32)}})
```

And vm peak across different onnxruntime versions:
    Testing onnxruntime-gpu 1.23.0 ... VmPeak = 3.6 GB
    Testing onnxruntime-gpu 1.23.2 ... VmPeak = 7.7 GB
    Testing onnxruntime-gpu 1.25.1 ... VmPeak = 7.7 GB

### Urgency

We are pinned to a version older than 1.23.2 and that avoids this issue.

### Platform

Windows

### OS Version

Windows 11

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.23.2

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

### Execution Provider Library Version

_No response_

### Model File

    tf.keras.backend.set_image_data_format("channels_last")
    x = tf.keras.Input((32, 64, 64, 1))
    m = tf.keras.Model(x, tf.keras.layers.Conv3D(6, 7, padding="same")(x))
    onnx.save(tf2onnx.convert.from_keras(m)[0], "{os.path.abspath(ONNX_FILE)}")

### Is this a quantized model?

Unknown

### PR DESCRIPTION
**Pull request by hariharans29**
### Description

https://github.com/microsoft/onnxruntime/pull/26103 which was shipped with ORT 1.23.2 introduced logic to parallelize on the batch and/or group count parameters for NCHW Conv if either were > 1 to improve performance of batched grouped convolutions - prior to that the parallelization was only on the output spatial map. The per-thread scratch buffer allocated by the B/G parallelization path introduced in https://github.com/microsoft/onnxruntime/pull/26103 is excessive and is not needed for the algorithm being used. Under the hood, `MlasConvOperation()` uses a double tiled approach which means that the per-thread scratch buffer needs just enough memory to fit that tile of data and nothing more than that. This change fixes the conservative over-allocation which shows up as a memory bloat in OP's model in issue https://github.com/microsoft/onnxruntime/issues/28466

Note: the work partitioning across (batch × group) introduced in #26103 is unchanged — only the per-thread scratch buffer sizing is corrected. The batched/grouped Conv speedup from #26103 is retained.

### Motivation and Context
Fix https://github.com/microsoft/onnxruntime/issues/28466

### CODE REVIEW

File: onnxruntime/test/contrib_ops/fused_conv_test.cc [hunk 1]
```diff
@@ -235,6 +235,54 @@ TEST(FusedConvTest, Cpu_Conv2D_Bias_Z_Relu) {
...
+                                   static_cast<int64_t>(input_width)};
+
+  vector<float> X(batch_count * input_channels * input_depth * input_height * input_width, 0.0f);
+  vector<float> W(filter_count * input_channels * kernel_depth * kernel_height * kernel_width, 0.0f);
+  vector<float> Y(batch_count * filter_count * input_depth * input_height * input_width, 0.0f);
+
+  test.AddInput<float>("X", X_shape, X);
+  test.AddInput<float>("W", W_shape, W, true);
+  test.AddOutput<float>("Y", Y_shape, Y);
+
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/fused_conv_test.cc:455]: Nitpick: This test uses all-zero inputs and weights, so the expected output is trivially all zeros. It validates that the code doesn't crash with the smaller buffer (and ASAN would catch OOB writes), but it doesn't exercise numerical correctness for non-trivial convolution on the batch/group path.

Consider using small non-zero constant values for `W` (e.g., `1.0f`) so the test verifies actual computation, not just that zeros propagate through without faulting.

**Review by tianleiwu — DISMISSED** [review]
## APPROVE

Clean, well-motivated fix for a significant memory over-allocation regression in the MLAS batch/group convolution path (introduced in #26103, shipped with ORT 1.23.2).

### Analysis

`MlasConvOperation` uses a double-tiled approach: it dynamically adjusts `StrideN` and `StrideK` (one doubles while the other halves), preserving the invariant `StrideN × StrideK = MLAS_SGEMM_STRIDEN × MLAS_SGEMM_STRIDEK = 16,384`. The im2col/vol2col functions write exactly `CountK × CountN ≤ StrideK × StrideN` elements to `ColumnBuffer`. This confirms each thread needs exactly `MLAS_CONV_WORKING_BUFFER_SIZE_PER_THREAD` — matching what the existing output-space threading path (`MlasConvOperationThreaded`) already uses.

The impact for the reported Conv3D case (InputShape=[32,64,64], Kernel=[7,7,7], BatchCount=12): old per-thread allocation was `OutputSize × K ≈ 45M floats ≈ 172 MB/thread` (over 2 GB total). The fix reduces this to 64 KB/thread (~768 KB total).

Both the allocation sizing and per-thread buffer indexing are updated consistently. Work partitioning across (batch × group) is untouched. Tests cover the fix from two angles: end-to-end crash/ASAN safety and direct buffer-size assertion.

One minor test improvement suggested inline.

File: onnxruntime/test/contrib_ops/fused_conv_test.cc (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/test/contrib_ops/fused_conv_test.cc:455]: Fixed.

**Review by tianleiwu — DISMISSED** [review]
No remaining findings on the current head. The buffer-size fix is consistent with `MlasConvOperation()`'s tiled `StrideN x StrideK` working-set invariant, and the earlier test concern was addressed by switching the new `FusedConv` regression to non-zero inputs with numerically meaningful expected values. Approving.

**Comment by hariharans29** [comment]
Will merge this after verifying a customer model perf on ARM server. Their model heavily relies on this path and I want to make sure its perf is not affected (in theory it should not be, but worth double checking)

**Review by tianleiwu — APPROVED** [review]

**Comment by hariharans29** [comment]
> Will merge this after verifying a customer model perf on ARM server. Their model heavily relies on this path and I want to make sure its perf is not affected (in theory it should not be, but worth double checking)

Double checked the perf numbers with this change. The perf stays the same. Merging.

**Comment by albertopasqualetto** [comment]
Hey @hariharans29 when will this code be released officially?

**Comment by hariharans29** [comment]
> Hey @hariharans29 when will this code be released officially?

It will be part of 1.27

**Comment by albertopasqualetto** [comment]
> > Hey @hariharans29 when will this code be released officially?
> 
> It will be part of 1.27

Is there an indicative release date? (The Web page is not updated)

**Comment by hariharans29** [comment]
> > > Hey @hariharans29 when will this code be released officially?
> > 
> > 
> > It will be part of 1.27
> 
> Is there an indicative release date? (The Web page is not updated)

I think 1.27 Python wheels should already be available on PyPi. If the nuget packages and/or other release artifacts are not yet available, it will be very soon (folks are working on it as we speak)

**Comment by albertopasqualetto** [comment]
> > > > Hey @hariharans29 when will this code be released officially?
> > > 
> > > 
> > > It will be part of 1.27
> > 
> > Is there an indicative release date? (The Web page is not updated)
> 
> I think 1.27 Python wheels should already be available on PyPi. If the nuget packages and/or other release artifacts are not yet available, it will be very soon (folks are working on it as we speak)
> 

You are right, it is on pypi. Thanks! I wasn't seeing the release on github.

**Comment by hariharans29** [comment]
> > > > > Hey @hariharans29 when will this code be released officially?
> > > > 
> > > > 
> > > > It will be part of 1.27
> > > 
> > > 
> > > Is there an indicative release date? (The Web page is not updated)
> > 
> > 
> > I think 1.27 Python wheels should already be available on PyPi. If the nuget packages and/or other release artifacts are not yet available, it will be very soon (folks are working on it as we speak)
> 
> You are right, it is on pypi. Thanks! I wasn't seeing the release on github.

We will officially announce it on GH once all the release artifacts are published and the entire release process is officially complete
